ddtrace/internal/flare.py

```diff
@@ -0,0 +1,184 @@
+                continue
+            args = c.get("args", {})
+
+            self._revert_configs()
+
+            # We only want the flare to be sent once, even if there are
+            # multiple tracer instances
+            if not os.path.exists(TRACER_FLARE_LOCK):
+                open(TRACER_FLARE_LOCK, "w").close()
+                data = {
+                    "case_id": args.get("case_id"),
+                    "source": "tracer_python",
+                    "hostname": args.get("hostname"),
+                    "email": args.get("user_handle"),
+                }
+                try:
+                    client = get_connection(config._trace_agent_url, timeout=self.timeout)
+                    headers, body = self._generate_payload(data)
+                    client.request("POST", TRACER_FLARE_ENDPOINT, body, headers)
+                    response = client.getresponse()
+                    if response.status == 200:
+                        log.info("Successfully sent the flare")
+                    elif response.status == 400:
+                        log.error(
+                            "Upload failed with 400 status code:(%s) %s", response.reason, response.read().decode()
+                        )
+                except Exception as e:
+                    raise Exception("Failed to send tracer flare: %s" % e)
+                finally:
+                    client.close()
+                    # Clean up files regardless of success/failure
+                    self._clean_up_files()
+
+    def _generate_config_file(self, pid: int):
+        config_file = f"{TRACER_FLARE_DIRECTORY}/tracer_config_{pid}.json"
+        try:
+            with open(config_file, "w") as f:
+                tracer_configs = {
+                    "configs": config.__dict__,
+                }
+                json.dump(
+                    tracer_configs,
+                    f,
+                    default=lambda obj: obj.__repr__() if hasattr(obj, "__repr__") else obj.__dict__,
+                    indent=4,
+                )
+        except Exception as e:
+            log.warning("Failed to generate %s: %s", config_file, e)
+            if os.path.exists(config_file):
+                os.remove(config_file)
+
+    def _revert_configs(self):
+        ddlogger = get_logger("ddtrace")
+        handler_to_remove = _get_handler(ddlogger, TRACER_FLARE_FILE_HANDLER_NAME)
+        if handler_to_remove:
+            ddlogger.removeHandler(handler_to_remove)
+            log.debug("ddtrace logs will not be routed to the %s file handler anymore", TRACER_FLARE_FILE_HANDLER_NAME)
+        else:
+            log.debug("Could not find %s to remove", TRACER_FLARE_FILE_HANDLER_NAME)
+        ddlogger.setLevel(self.original_log_level)
+
+    def _generate_payload(self, params: Dict[str, str]) -> Tuple[dict, bytes]:
+        tar_stream = io.BytesIO()
+        with tarfile.open(fileobj=tar_stream, mode="w") as tar:
```

Comment:
how big does this usually end up being? do we have any concerns over holding this in memory?

if it is like <5mb, ignore me. if we think it could grow 20+mb, I wonder if we'll have memory usage issues?

I am probably making mountain from a mole hill, but thought I'd check your thoughts.

Reply:
We do limit individual files to 15mb by default, but you're right we don't have a bound on how big the directory can become. One idea off the top of my head is we check the size of the directory in its current state, and if it's above some threshold we set, we don't add more flare files to it. This would be further leaning on the "best effort" nature of the tracer flare, but would be pretty easy to implement.

Reply:
when you iterate over the files, start with the configs, then add as many log files up until you hit the limit?

but I am wondering if there is a way to include that log files were skipped due to size.... or maybe you take them pid by pid, so you always have config + log for a given pid... 🤔

Reply:
Yeah I think I have to do it in the pairs. If anything, the log files will already be there anyways when I add the file handler to the logger... So the reverse would be more doable